src/backend-c/runtime/hashcons.c

```diff
@@ -48,46 +48,44 @@ int64_t hashcons(table ht, int64_t e, int64_t hcode)
     new_item->next = NULL;
     C->list = new_item;
     ht->num_elems++;
-    return e;
+    return (int64_t) e;
   }
   list p = C->list;
   while (p != NULL) {
     if (types_equal(e, p->data))
-      return p->data;
+      return (int64_t) p->data;
     p = p->next;
   }
   list new_item = GC_MALLOC(8 * 2);
   new_item->data = e;
   new_item->next = C->list;
   C->list = new_item;
   ht->num_elems++;
-  return e;
+  return (int64_t) e;
 }
 
-int types_equal(int64_t t1, int64_t t2)
+int types_equal(int64_t* t1, int64_t* t2)
 {
-  int64_t tag1 = (t1 & 7);
-  int64_t tag2 = (t2 & 7);
-  int64_t count, untagged_t1, untagged_t2, i;
+  int64_t tag1 = t1[0];
+  int64_t tag2 = t2[0];
+  int64_t count, i;
   if (tag1 == tag2) {
-    untagged_t1 = (t1 ^ tag1);
-    untagged_t2 = (t2 ^ tag1);
     switch (tag1) {
```

Comment:
This case uses a lot of magic numbers and would benefit with some comments that describe what those magic numbers mean. I know this isn't really a product of this PR, but I just realized while reading the code that I wasn't able to figure out the mapping easily.

Reply:
I will open another PR now to have this sorted out.

Reply:
https://github.com/Gradual-Typing/Grift/pull/29

Reply:
Done.